feat(editor): Allow users to update verified nodes from the node settings panel

dana-gill opened the pull request:
## Summary

Allow users to update verified nodes from the node settings panel

This PR also does the following:
- Renames `activePackage` to `unverifiedNodesPackage` to discern between verified community node and unverified community nodes as unverified community nodes comes from Community Node store but verified community nodes comes from Node Types store
- Includes `semver` as dependency to differentiate between the currently installed version and version available in strapi. We already have this dependency installed in the root and we need to make it accessible to the FE
- Includes a notice for users who enable both verified and unverified nodes since we want to prevent them from installing verified nodes with versions we haven't vetted yet

To test with verified nodes and Verdaccio:
1. Cherry-pick `db80abc4fd7057554983f1cfd1ebb20e08f0081e`
2. Set `export N8N_UNVERIFIED_PACKAGES_ENABLED=false` and `export N8N_VERIFIED_PACKAGES_ENABLED=true`
3. Follow the instructions in the Verdaccio doc on Notion

To test with verified nodes and npm:
1. Set `export N8N_UNVERIFIED_PACKAGES_ENABLED=false` and `export N8N_VERIFIED_PACKAGES_ENABLED=true`

To test with verified and unverified nodes:
1. Set `export N8N_UNVERIFIED_PACKAGES_ENABLED=true` and `export N8N_VERIFIED_PACKAGES_ENABLED=true`

To test with unverified nodes only:
1. Set `export N8N_UNVERIFIED_PACKAGES_ENABLED=true` and `export N8N_VERIFIED_PACKAGES_ENABLED=false`

## Related Linear tickets, Github issues, and Community forum posts

https://linear.app/n8n/issue/NODE-3049/when-new-approved-version-of-existing-verified-node-is-approved-allow

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

Commit:
Allow for updating a community node via settings

Commit:
add tests

Commit:
Add warning when users are using both verified and unverified comm nodes

Commit:
manage copy for updating version

Commit:
refactor isVerifiedLatestPackage and packageVersion

Commit:
make typecheck happy

Commit:
add more tests

Commit:
make variable more accurate

dana-gill committed:
Update packages/frontend/editor-ui/package.json

Co-authored-by: cubic-dev-ai[bot] <191113872+cubic-dev-ai[bot]@users.noreply.github.com>

michael-radency commented on packages/cli/src/controllers/community-packages.controller.ts:
```diff
@@ -264,6 +264,7 @@ export class CommunityPackagesController {
 			const newInstalledPackage = await this.communityPackagesService.updatePackage(
 				this.communityPackagesService.parseNpmPackageName(name).packageName,
 				previouslyInstalledPackage,
+				version,
```
this is correct
we either need to respect version sent or remove this option as it's not used
![image](https://github.com/user-attachments/assets/2c83ebc6-ed87-46fc-9565-f9085fc274e0)

michael-radency commented on packages/cli/src/services/community-packages.service.ts:
```diff
@@ -349,8 +349,9 @@ export class CommunityPackagesService {
+		version?: string,
 	): Promise<InstalledPackages> {
-		return await this.installOrUpdatePackage(packageName, { installedPackage });
+		return await this.installOrUpdatePackage(packageName, { installedPackage, version });
```
needs to be updated here
![image](https://github.com/user-attachments/assets/23ca14d7-85dc-49f4-a914-92a6b434400b)

michael-radency commented on packages/frontend/editor-ui/src/components/CommunityPackageCard.vue:
```diff
@@ -56,6 +75,24 @@ function onUpdateClick() {
 	openCommunityPackageUpdateConfirmModal(props.communityPackage.packageName);
 }
+
+watch(
+	() => props.communityPackage?.packageName,
+	async (packageName) => {
+		if (packageName) {
+			await nodeTypesStore.loadNodeTypesIfNotLoaded();
+			const nodeType = nodeTypesStore.visibleNodeTypes.find((node) =>
+				node.name.includes(packageName),
+			);
+
+			const attributes = await nodeTypesStore.getCommunityNodeAttributes(nodeType?.name || '');
+			if (attributes?.npmVersion) {
+				latestVerifiedVersion.value = attributes.npmVersion;
+			}
+		}
+	},
+	{ immediate: true },
+);
```
why do we need this in `watch`?

dana-gill replied:
I needed to get this in a watch since otherwise the `Update` button does not disappear when the update is successfully finished. The update for verified community nodes has to follow a slightly different path in comparison to unverified community nodes

michael-radency replied:
you probably don't need to do this, it will be handled in the higher component
![image](https://github.com/user-attachments/assets/781093c3-45b6-44c8-aa7b-03c724c9b3c3)
maybe try to set --registry here
![image](https://github.com/user-attachments/assets/4b101434-88b6-45c1-8a99-716add2e71cf)

dana-gill replied:
https://github.com/n8n-io/n8n/pull/16447/commits/0b7ea569dc5b075b7948f0c6183c79f0f81b7126

michael-radency commented on packages/cli/src/services/community-packages.service.ts:
```diff
@@ -349,8 +349,9 @@ export class CommunityPackagesService {
+		version?: string,
 	): Promise<InstalledPackages> {
-		return await this.installOrUpdatePackage(packageName, { installedPackage });
+		return await this.installOrUpdatePackage(packageName, { installedPackage, version });
```
should we pass checksum here to validate?

Commit:
update version

Commit:
add checksum

Commit:
Add test for checksum

Commit:
add semver types

michael-radency commented on packages/cli/src/services/community-packages.service.ts:
```diff
@@ -386,10 +388,12 @@ export class CommunityPackagesService {
-		const packageVersion = isUpdate || !options.version ? 'latest' : options.version;
+		const packageVersion = !options.version ? 'latest' : options.version;
 
 		const shouldValidateChecksum = 'checksum' in options && Boolean(options.checksum);
 		this.checkInstallPermissions(isUpdate, shouldValidateChecksum);
```
`checkInstallPermissions` has to be updated to work with update
```
	private checkInstallPermissions(isUpdate: boolean, checksumProvided: boolean) {
		if (isUpdate) return;

		if (!this.globalConfig.nodes.communityPackages.unverifiedEnabled && !checksumProvided) {
			throw new UnexpectedError('Installation of unverified community packages is forbidden!');
		}
	}
```

michael-radency commented on packages/cli/src/services/community-packages.service.ts:
```diff
@@ -386,10 +388,12 @@ export class CommunityPackagesService {
 		packageName: string,
-		options: { version?: string; checksum?: string } | { installedPackage: InstalledPackages },
+		options:
+			| { version?: string; checksum?: string }
+			| { installedPackage: InstalledPackages; version?: string; checksum?: string } = {},
```
checksum will not be validated if update
```
		if (!isUpdate && options.checksum) {
			await verifyIntegrity(packageName, packageVersion, this.getNpmRegistry(), options.checksum);
		}
```

michael-radency commented on packages/frontend/editor-ui/src/components/CommunityPackageManageConfirmModal.vue:
```diff
@@ -107,16 +123,31 @@ const onUninstall = async () => {
+
+		if (settingsStore.isCommunityNodesFeatureEnabled) {
+			updatedVersion = nodeTypeStorePackage.value?.npmVersion;
+			await communityNodesStore.updatePackage(
+				props.activePackageName,
+				updatedVersion,
+				nodeTypeStorePackage.value?.checksum,
+			);
```
updating to version higher then verified not possible on self-hosted, no changes to env defaults

dana-gill replied:
This should be handled by https://arc.net/l/quote/ksncvslg

Commit:
flip logic for unverified and verified

this way when all community nodes are enabled, the unverified path takes
precedence

Commit:
validate checksum for update

Commit:
fix typecheck

michael-radency committed:
add registry to get all packages

michael-radency committed:
Display correct version for update when unverified nodes disabled

dana-gill commented on packages/cli/src/controllers/community-packages.controller.ts:
```diff
@@ -164,7 +164,12 @@ export class CommunityPackagesController {
 
 		try {
-			const command = ['npm', 'outdated', '--json'].join(' ');
+			const command = [
+				'npm',
+				'outdated',
+				'--json',
+				`--registry=${this.communityPackagesService.getNpmRegistry()}`,
+			].join(' ');
```
bug fix (technically)

dana-gill commented on packages/frontend/editor-ui/src/components/CommunityPackageCard.vue:
```diff
@@ -135,10 +97,7 @@ watch(
-					v-else-if="hasUnverifiedPackagesUpdate || hasVerifiedPackageUpdate"
-					placement="top"
-				>
+				<n8n-tooltip v-else-if="communityPackage.updateAvailable" placement="top">
```
Notes from pairing with @michael-radency: We need to check if this can work with verdaccio. It could be solved by the addition of `registry` in controller with this commit

dana-gill commented on packages/frontend/editor-ui/src/components/CommunityPackageCard.vue:
```diff
@@ -76,24 +56,6 @@ function onUpdateClick() {
 	openCommunityPackageUpdateConfirmModal(props.communityPackage.packageName);
 }
-
-watch(
```
Notes from pairing with @michael-radency: May not need watch because we could check:
1. check for package version based on registry in the controller change
2. onlyAllowVerifiedPackages=true, selfhosted=true => check if this still works without the `watch`
3. show update only if versionInStrapi (aka nodeTypeStore) >= currentInstalledPackage

But I did try this and it didn't work :(

Concern of michael: Want to use current implementation and things should be handled in BE not FE

Commit:
checkout pnpm-lock to master

Commit:
update lock

Commit:
Revert "add registry to get all packages"

This reverts commit d216da176402d0cfbf3a5ba79fdf232aa9d6e8e2.

michael-radency reviewed (approved):
LGTM :+1:
